File: go/instruction_serialiser/modulo_node.go
package instruction_serialiser

import "math"

func (n *ModuloNode) calculate(parameters map[string]interface{}) (ArithmeticType, error) {
	var leftVal ArithmeticType
	{
		var err error
		if leftVal, err = n.GetLeftChild().Calculate(parameters); err != nil {
			return 0, err
		}
	}
	var rightVal ArithmeticType
	{
		var err error
		if rightVal, err = n.GetRightChild().Calculate(parameters); err != nil {
			return 0, err
		}
	}
	return math.Mod(leftVal, rightVal), nil
}

func (a *ArithmeticResultNodeWrapper) Modulo(divisorNode *ArithmeticResultNodeWrapper) *ArithmeticResultNodeWrapper {
	return &ArithmeticResultNodeWrapper{
		Node: &ArithmeticResultNodeWrapper_ModuloNode{
			ModuloNode: &ModuloNode{
				LeftChild:  a,
				RightChild: divisorNode,
			},
		},
	}
}

File: typescript/src/modulo_node.ts
import { CalculatableNode } from './calculatable_node';
import { ModuloNode } from './instruction_serialiser/instruction_serialiser_pb';
import { InstructionParameters } from './instruction_parameters';

declare module './instruction_serialiser/instruction_serialiser_pb' {
  interface ModuloNode extends CalculatableNode<number> {}
}

ModuloNode.prototype.calculate = function(parameters: InstructionParameters): number {
  return this.getLeftChild().calculate(parameters) %
    this.getRightChild().calculate(parameters);
};

export { ModuloNode };
